core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -6069,7 +6110,13 @@ public SqlNode expandGroupByOrHavingExpr(SqlNode expr,
             }
             return param;
           }
+          @Override public SqlNode visit(SqlNamedParam param) {
```

Comment:
Do we need the f (alreadyVisited.add(param)) { ... } present in the dynamicParam case?

Reply:
No. That arises because DynamicParams are just ? and they get assigned an index, so the visiting order matters. Our order doesn't matter so we don't need to support that.